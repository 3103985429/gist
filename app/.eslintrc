{
  // extend root arrays
  "extends": "../.eslintrc.js",
  "parserOptions": {
    "ecmaVersion": 6,
  },
  "env": {
    "browser": false,
    "node": true
  }
}
