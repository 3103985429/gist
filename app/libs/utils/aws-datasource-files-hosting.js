const winston = require('winston');
const aws = require('aws-sdk');
const fs = require('fs');
const _ = require('lodash');

const bucket = process.env.DO_S3_BUCKET;
const endpoint = new aws.Endpoint(process.env.DO_S3_ENDPOINT);
const s3 = new aws.S3({
    endpoint,
    accessKeyId: process.env.DO_ACCESS_KEY_ID,
    secretAccessKey: process.env.DO_SECRET_ACCESS_KEY,
});
const debugS3 = false;

function _uploadAncillaryFile(filePath, fileName, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getAncillaryAssetKey(teamSubdomin, datasetId, fileName);
    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the ancillary file to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadAncillaryFile = _uploadAncillaryFile;

function _copyAncillaryFile(team_subdomain, originId, destinationId, fileName, callback) {
    const key = getAncillaryAssetKey(team_subdomain, destinationId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getAncillaryAssetKey(team_subdomain, originId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the ancillary file from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyAncillaryFile = _copyAncillaryFile;

function _uploadDataSource(filePath, newFilename, contentType, teamSubdomin, datasetId, callback) {
    const file = fs.createReadStream(filePath);
    const key = getDatasourceKey(teamSubdomin, datasetId, newFilename);

    const params = {
        Bucket: bucket,
        Key: key,
        ContentType: contentType,
        Body: file,
        ACL: 'private',
    };

    debug(`Uploading the datasource to S3: ${key}`);

    s3.upload(params, function (err) {
        if (err) {
            return callback(err);
        }

        callback();
    });
}

module.exports.uploadDataSource = _uploadDataSource;

function _copySampleDatasource(datasetToDuplicateId, fileName, datasetId, teamSubdomain, callback) {
    const key = getDatasourceKey(teamSubdomain, datasetId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: getDatasourceKey('sampleteam', datasetToDuplicateId, fileName, bucket),
        Key: key,
    };

    debug(`Copying the sample datasource from S3: ${key}`);

    s3.copyObject(params, function (err) {
        if (err) {
            winston.error(`S3: error copying sample datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copySampleDatasource = _copySampleDatasource;

function _copyDatasource(team_subdomain, fileName, copyFromId, copyToId, callback) {
    const keyFrom = getDatasourceKey(team_subdomain, copyFromId, fileName, bucket);
    const keyTo = getDatasourceKey(team_subdomain, copyToId, fileName);
    const params = {
        Bucket: bucket,
        CopySource: keyFrom,
        Key: keyTo,
    };

    debug(`Copying the datasource from S3: from ${keyFrom} to ${keyTo}`);

    s3.copyObject(params, function (err, data) {
        if (err) {
            winston.error(`S3: error copying datasource: ${err.message}`);
            return callback(err);
        }

        callback();
    });
}

module.exports.copyDatasource = _copyDatasource;

function _getDatasource(description) {
    const key = getDatasourceKey(description._team.subdomain, description.schemaId || description._id, description.fileName);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading the datasource from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getDatasource = _getDatasource;

function _getAncillaryFile(description) {
    const key = getAncillaryAssetKey(description._team.subdomain, description._id, description.ancillaryFile);

    const param = {
        Bucket: bucket,
        Key: key,
    };
    debug(`Reading ancillary file from S3: ${key}`);

    return s3.getObject(param);
}

module.exports.getAncillaryFile = _getAncillaryFile;

function keyExists(description, callback = _.noop) {
    return new Promise((resolve, reject) => {
        const key = getDatasourceKey(description._team.subdomain, description.schema_id || description._id, description.fileName);

        const params = {
            Bucket: bucket,
            Key: key,
        };
        debug(`Verifying the datasource from S3: ${key}`);

        s3.headObject(params, (err) => {
            if (err) {
                reject(err);
                return callback(err, false);
            }

            resolve();
            callback(null, true);
        });
    });
}

module.exports.keyExists = keyExists;

function _deleteDataset(description, callback) {
    const team_subdomain = description._team.subdomain;
    const key = getDatasetKey(team_subdomain, description.id);
    const param = {
        Bucket: bucket,
        Prefix: key,
    };

    debug(`Removing the datasource from S3: ${key}`);

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Deleted) {
                return callback();
            }

            if (data.Deleted.length === 1000) {
                return _deleteDataset(description, callback);
            }

            callback();
        });
    });
}

module.exports.deleteDataset = _deleteDataset;

function _deleteTeam(subdomain, callback) {
    const param = {
        Bucket: bucket,
        Prefix: getTeamKey(subdomain),
    };

    s3.listObjects(param, function (err, data) {
        if (err) {
            return callback(err);
        }

        if (!data.Contents) {
            return callback();
        }

        if (data.Contents.length === 0) {
            return callback();
        }

        const param = { Bucket: bucket };
        param.Delete = { Objects: [] };

        data.Contents.forEach(function (content) {
            param.Delete.Objects.push({ Key: content.Key });
        });

        s3.deleteObjects(param, function (err, data) {
            if (err) {
                return callback(err);
            }

            if (!data.Contents) {
                return callback();
            }

            if (data.Contents.length === 1000) {
                return _deleteTeam(subdomain, callback);
            }

            callback();
        });
    });
}

module.exports.deleteTeam = _deleteTeam;

function _deleteObject(key, callback) {
    const params = {
        Bucket: bucket,
        Key: key,
    };

    s3.deleteObject(params, function (err, data) {
        if (err) {
            return callback(err);
        }

        callback(null, data);
    });
}

module.exports.deleteObject = _deleteObject;

function getDatasourceKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `datasources/${fileName}`, bucket);
}

function getAncillaryAssetKey(subdomain, datasetId, fileName, bucket = null) {
    return getDatasetKey(subdomain, datasetId, `assets/ancillary/${fileName}`, bucket);
}

function getDatasetKey(subdomain, datasetId, resourceName = '', bucket = null) {
    return getTeamKey(subdomain, `datasets/${datasetId}/${resourceName}`, bucket);
}

function getTeamKey(subdomain, resourceName = '', bucket = null) {
    const prefix = bucket ? `${bucket}/` : '';

    return `${prefix}${subdomain}/${resourceName}`;
}

function debug(msg) {
    if (debugS3) {
        winston.debug(msg);
    }
}
