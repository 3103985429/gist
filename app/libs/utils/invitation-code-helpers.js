const crypto = require('crypto');

/**
 * Returns a sha512 hash of the given string.
 * @param confirmation_code - Code to hash
 * @return
 */
module.exports.createHashWithString = confirmation_code => {
    // Create a new hash, and then update it and return the digest
    const hash = crypto.createHash('sha512');
    hash.update(confirmation_code);
    return hash.digest('hex');
};

/**
 * Generates a random short code.
 * @param codeSize - Size in bytes to generate a code
 * @return
 */
module.exports.generateRandomCode = codeSize => {
    // Get a random buffer
    const randomBuffer = crypto.randomBytes(codeSize);
    // Return the random buffer to string
    return randomBuffer.toString('hex');
};
