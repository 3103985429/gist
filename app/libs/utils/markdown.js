var fs = require('fs');
var marked = require('marked');

module.exports.getMarkdown = function (filePath) {
    var rawMarkdown = fs.readFileSync(filePath, 'utf-8');
    return rawMarkdown;
};

module.exports.parseMarkdown = function (markdown) {
    var options = {
        renderer: new marked.Renderer(),
        gfm: true,
        tables: true,
        breaks: false,
        pedantic: false,
        sanitize: true,
        smartLists: true,
        smartypants: false
    };
    marked.setOptions(options);

    var lexer = new marked.Lexer(options);
    var parsedBody = marked(markdown);
    return parsedBody;
};

module.exports.replaceCharts = function (parsed) {
    var chartRE = /((?:<p>)?&lt;chart data-source-key=&quot;(?=\S*)([a-zA-Z\d_-]*)&quot; data-chart-type=&quot;(?=\S*)([a-zA-Z\d-]*)&quot; class=&quot;chart&quot; data-revision=&quot;(\d+)&quot; \/&gt;(?:<\/p>)?)/g;
    var parsedReplaced = parsed.replace(chartRE, '<div id="$2" data-source-key="$2" data-chart-type="$3" class="$3 chart" data-revision="$4"></div>');
    return parsedReplaced;
};
