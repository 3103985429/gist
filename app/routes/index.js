const { defaultTo } = require('lodash');
const helmet = require('helmet');
const express = require('express');

const { errorLogger } = require('../utils/winston');
const { checkCors, API_ALLOWED } = require('./permissions');
const { version } = require('../../package.json');

const PROTOCOL = process.env.USE_SSL === 'true' ? 'https' : 'http';
const HOST = defaultTo(process.env.HOST, 'localhost:9080');
const ROOT_DOMAIN = `${PROTOCOL}://app.${HOST}`;
const API_VERSION = 'v1';

const _mountRoutes_monitoring = function (app) {
    app.get('/_ah/health', function (req, res) {
        res.set('Content-Type', 'text/plain');
        res.status(200).send('ok');
    });
};

const _mountRoutes_errorHandling = function (app) {
    // Add the error logger after all middleware and routes so that
    // it can log errors from the whole application. Any custom error
    // handlers should go after this.
    errorLogger(app);

    // eslint-disable-next-line no-unused-vars
    app.use(function (err, req, res, next) {
        const status = err.status || 404;
        const [title, message] = (status === 404) ?
            ['Oops!', 'The page you are looking for doesn\'t exist or has been moved.'] :
            ['Uh oh.', 'Something went wrong\u2026'];
        const error = process.env.NODE_ENV === 'development' ? err.toString() : `Error: ${status}`;

        res.status(status).render('error', {
            env: process.env,
            title,
            message,
            err: error,
            has_google_analytics: !!process.env.GOOGLE_ANALYTICS_ID,
        });
    });

    app.use(function (req, res) {
        res.status(404);
        res.render('error', {
            env: process.env,
            title: 'Oops!',
            message: 'The page you are looking for doesn\'t exist or has been moved.',
            err: 'Error: 404',
            has_google_analytics: !!process.env.GOOGLE_ANALYTICS_ID,
        });
    });
};

const _mountRoutes_endPoints = function (app) {
    app.all('*', function ({ subdomains, url }, res, next) {
        const redirectToRootDomainPatterns = [
            /^\/dashboard/,
            /^\/login/,
            /^\/signup/,
            /^\/auth/,
            /^\/product/,
            /^\/news/,
            /^\/pricing/,
            /^\/plans/,
        ];
        const isEnterprise = process.env.NODE_ENV === 'enterprise';
        const isRootDomain = subdomains.length === 1 && ['app', 'www'].includes(subdomains[0]);

        if (!isEnterprise && !isRootDomain && redirectToRootDomainPatterns.some(pattern => pattern.test(url))) {
            return res.redirect(ROOT_DOMAIN + url);
        }

        next();
    });

    app.use('/robots.txt', express.static(__dirname + '/../../robots.txt'));

    app.use('/', require('./homepage'));

    app.use('/s', require('./shared_pages'));
    app.use('/presentation', require('./stories'));
    app.use(`/json-api/${API_VERSION}`, checkCors(API_ALLOWED), function (req, res, next) {
        req.apiVersion = API_VERSION;
        next();
    }, require('./json-api/router'));
    app.use('/auth', helmet.frameguard(), checkCors(), require('./auth'));
    app.use('/login', helmet.frameguard(), checkCors(), function (req, res) {
        res.redirect('/auth/login');
    });
    app.use('/signup', helmet.frameguard(), checkCors(), require('./signup'));
    app.use('/reset', helmet.frameguard(), checkCors(), function (req, res) {
        // TODO: Perform an jwt auth here instead of in account.js / or better yet auth in both!
        res.render('auth/password', {
            env: process.env,
        });
    });

    app.use('/dashboard', helmet.frameguard(), checkCors(), require('./dashboard'));

    app.use('/api', helmet.frameguard(), require('./api'));

    app.use('/account', checkCors(), require('./account'));
    app.use('/edit-visualization', require('./external/editVisualization'));
    app.use('/create-visualization', require('./external/createVisualization'));
    app.use('/visualization-list', require('./external/visualizationList'));

    app.use('/', require('./articles'));
    app.use('/', checkCors(API_ALLOWED), require('./views').router);
};

module.exports.MountRoutes = function (app) {
    app.get('/env', function (req, res) {
        const obj = {
            node_env: process.env.NODE_ENV,
            host: HOST,
            s3Bucket: process.env.DO_S3_BUCKET,
            s3Domain: process.env.DO_S3_ENDPOINT,
            subdomain: process.env.SUBDOMAIN,
            authProtocol: process.env.AUTH_PROTOCOL,
            hasSubteams: process.env.SUBTEAMS,
            git: {
                branch: process.env.BRANCH_NAME,
                commit: process.env.COMMIT_HASH,
            },
            version,
            insight: process.env.INSIGHT,
            hide_business_links: process.env.HIDE_BUSINESS_LINKS,
            hasSmartsheetInfo: !!process.env.SMARTSHEET_CLIENT_ID,
            hasPipedriveInfo: !!process.env.PIPEDRIVE_CLIENT_ID,
            hasDatadotworldInfo: !!process.env.DATA_DOT_WORLD_CLIENT_ID,
            hasSalesforceInfo: !!process.env.SALESFORCE_CONSUMER_KEY
        };

        return res.json(obj);
    });

    _mountRoutes_monitoring(app);
    _mountRoutes_endPoints(app);
    _mountRoutes_errorHandling(app);
};
