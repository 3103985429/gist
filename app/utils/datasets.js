const { findDataset } = require('../controllers/api/dataset/get-helpers');
const { some, includes } = require('lodash');
const DATA_SOURCE_TYPES = require('../../config/data-source-types');

module.exports.getDataset = getDataset;
module.exports.hasValidDatasetSource = hasValidDatasetSource;

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getDataset(id) {
    const { description } = await findDataset({ id });

    return description;
}

/**
 * Determins if the datasourceDescription passed has a valid dataset source
 * determined from the list in config/data-source-types.js
 * @param {Object} datasourceDescription
 * @returns {Boolean}
 */
function hasValidDatasetSource(datasourceDescription) {
    return some(datasourceDescription, (value, key) => {
        return includes(DATA_SOURCE_TYPES, key);
    });
}
