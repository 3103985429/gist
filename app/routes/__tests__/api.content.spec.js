import request from 'supertest';

import app from '../../../app';
import { Lazy_Shared_ProcessedRowObject_MongooseContext } from '../../models/processed_row_objects';
import cached_values from '../../models/cached_values';
import Descriptions from '../../models/descriptions';
import { hydrateDb, getId } from '../../../internals/testing/backend/utils';
import { pokemon1 } from '../../../internals/testing/backend/fixtures/datasets';
import {
    pokemon1RowObject1,
    pokemon1RowObject5,
    pokemon1RowObject9,
    pokemon1RowObject7Unpublished,
} from '../../../internals/testing/backend/fixtures/pokemon1ProcessedRowObjects';
import {
    pokemon1RowObject10,
} from '../../../internals/testing/backend/fixtures/notInserted_pokemon1ProcessedRowObjects';
import { pokemon1CachedValues } from '../../../internals/testing/backend/fixtures/cachedValues';

const pokemon1Id = getId(pokemon1);
const pokemonNoField = 'Pokemon No_';
const pokemonType1Field = 'Type 1';
const pokemonMaxCPField = 'Max CP';


describe('Content API', () => {
    let appRequest, ProcessedRowsModel;

    beforeEach(async () => {
        await hydrateDb();
        appRequest = request(app);
        ProcessedRowsModel = Lazy_Shared_ProcessedRowObject_MongooseContext(pokemon1Id).Model;
    });

    describe('Publish processed row object', () => {
        it('should return proper response', async () => {
            const response = await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject7Unpublished)}`)
                .send({ published: true });

            expect(response).toMatchSnapshot();
        });

        it('should update published key in db', async () => {
            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject7Unpublished)}`)
                .send({ published: true });

            expect(await ProcessedRowsModel.findById(getId(pokemon1RowObject7Unpublished)))
                .toHaveProperty('published', true);
        });

        it('should insert published row\'s field values to cache', async () => {
            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject7Unpublished)}`)
                .send({ published: true });

            const cachedValue = await cached_values.findById(getId(pokemon1CachedValues));

            expect(cachedValue.limitedUniqValsByColName[pokemonNoField])
                .toContain(pokemon1RowObject7Unpublished.rowParams[pokemonNoField]);
        });

        it('should update dataset modify date', async () => {
            const { updatedAt: prevUpdatedAt } = await Descriptions.findById(pokemon1Id);

            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject7Unpublished)}`)
                .send({ published: true });

            const { updatedAt: nextUpdatedAt } = await Descriptions.findById(pokemon1Id);

            expect(nextUpdatedAt.valueOf()).toBeGreaterThan(prevUpdatedAt.valueOf());
        });
    });

    describe('Remove processed row object', () => {
        it('should return proper response', async () => {
            const response = await appRequest
                .delete(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`);
            const removedRow = await ProcessedRowsModel.findById(getId(pokemon1RowObject1));

            expect(response).toMatchSnapshot();
            expect(removedRow).toBeNull();
        });

        it('should remove object from db', async () => {
            await appRequest
                .delete(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`);

            expect(await ProcessedRowsModel.findById(getId(pokemon1RowObject1))).toBeNull();
        });

        it('should remove deleted row\'s cached field values if there is no more row objects with such values', async () => {
            await appRequest
                .delete(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`);

            const cachedValue = await cached_values.findById(getId(pokemon1CachedValues));

            expect(cachedValue.limitedUniqValsByColName[pokemonNoField])
                .not.toContain(pokemon1RowObject1.rowParams[pokemonNoField]);

            expect(cachedValue.limitedUniqValsByColName[pokemonMaxCPField])
                .not.toContain(pokemon1RowObject1.rowParams[pokemonMaxCPField]);
        });

        it('should remove deleted row\'s cached field values if there are only unpublished rows with those values', async () => {
            await appRequest
                .delete(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject9)}`);

            const cachedValue = await cached_values.findById(getId(pokemon1CachedValues));

            expect(cachedValue.limitedUniqValsByColName[pokemonNoField])
                .not.toContain(pokemon1RowObject9.rowParams[pokemonNoField]);

            expect(cachedValue.limitedUniqValsByColName[pokemonType1Field])
                .not.toContain(pokemon1RowObject9.rowParams[pokemonType1Field]);
        });

        it('should not remove deleted row\'s cached field values if there are still objects with those values', async () => {
            await appRequest
                .delete(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject5)}`);

            const cachedValue = await cached_values.findById(getId(pokemon1CachedValues));

            expect(cachedValue.limitedUniqValsByColName[pokemonNoField])
                .not.toContain(pokemon1RowObject5.rowParams[pokemonNoField]);

            expect(cachedValue.limitedUniqValsByColName[pokemonType1Field])
                .toContain(pokemon1RowObject5.rowParams[pokemonType1Field]);
        });

        it('should update dataset modify date', async () => {
            const { updatedAt: prevUpdatedAt } = await Descriptions.findById(pokemon1Id);

            await appRequest
                .delete(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject5)}`);

            const { updatedAt: nextUpdatedAt } = await Descriptions.findById(pokemon1Id);

            expect(nextUpdatedAt.valueOf()).toBeGreaterThan(prevUpdatedAt.valueOf());
        });
    });

    describe('Create processed row object', () => {
        it('should return proper response', async () => {
            const response = await appRequest
                .post(`/api/dataset/content/${pokemon1Id}`)
                .send(pokemon1RowObject10);

            expect(response.statusCode).toBe(200);

            // this cannot be stored in snapshot as it's random for each request
            delete response.body.pKey;
            expect(response).toMatchSnapshot();
        });

        it('should create proper object in db', async () => {
            await appRequest
                .post(`/api/dataset/content/${pokemon1Id}`)
                .send(pokemon1RowObject10);

            expect(await ProcessedRowsModel.findById(getId(pokemon1RowObject10)))
                .toMatchObject({
                    ...pokemon1RowObject10,
                    _id: expect.anything(),
                    pKey: expect.anything(),
                });
        });

        it('should insert created row\'s field values to cache', async () => {
            await appRequest
                .post(`/api/dataset/content/${pokemon1Id}`)
                .send(pokemon1RowObject10);

            const cachedValue = await cached_values.findById(getId(pokemon1CachedValues));

            expect(cachedValue.limitedUniqValsByColName).toMatchObject({
                [pokemonNoField]: expect.arrayContaining([10]),
                [pokemonType1Field]: expect.arrayContaining(['Bug']),
            });
        });

        it('should update dataset modify date', async () => {
            const { updatedAt: prevUpdatedAt } = await Descriptions.findById(pokemon1Id);

            await appRequest
                .post(`/api/dataset/content/${pokemon1Id}`)
                .send(pokemon1RowObject10);

            const { updatedAt: nextUpdatedAt } = await Descriptions.findById(pokemon1Id);

            expect(nextUpdatedAt.valueOf()).toBeGreaterThan(prevUpdatedAt.valueOf());
        });
    });

    describe('Update row object\'s params', () => {
        it('should return proper response', async () => {
            const response = await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`)
                .send({ rowParams: { [pokemonNoField]: 12 } });

            expect(response).toMatchSnapshot();
        });

        it('should update object in db', async () => {
            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`)
                .send({ rowParams: { [pokemonNoField]: 12 } });

            expect(await ProcessedRowsModel.findById(getId(pokemon1RowObject1)))
                .toHaveProperty(['rowParams', pokemonNoField], 12);
        });

        it('should add new values to cache', async () => {
            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`)
                .send({ rowParams: { [pokemonNoField]: 12, [pokemonType1Field]: 'SomethingNew' } });
            const cachedValue = await cached_values.findById(getId(pokemon1CachedValues));

            expect(cachedValue.limitedUniqValsByColName).toMatchObject({
                [pokemonNoField]: expect.arrayContaining([12]),
                [pokemonType1Field]: expect.arrayContaining(['SomethingNew']),
            });
        });

        it('should not add duplicate values to cache', async () => {
            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`)
                .send({ rowParams: { [pokemonNoField]: 5, [pokemonType1Field]: 'Fire' } });
            const cachedValue = await cached_values.findById(getId(pokemon1CachedValues));

            expect(cachedValue.limitedUniqValsByColName).toMatchObject({
                [pokemonType1Field]: ['Fire', 'Grass', 'Water'],
            });
        });

        it('should remove unique values from cache', async () => {
            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`)
                .send({ rowParams: { [pokemonNoField]: 5, [pokemonType1Field]: 'Fire' } });
            const cachedValue = await cached_values.findById(getId(pokemon1CachedValues));

            expect(cachedValue.limitedUniqValsByColName).toMatchObject({
                [pokemonNoField]: [2, 3, 4, 5, 9],
            });
        });

        it('should update dataset modify date', async () => {
            const { updatedAt: prevUpdatedAt } = await Descriptions.findById(pokemon1Id);

            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`)
                .send({ rowParams: { [pokemonNoField]: 10 } });

            const { updatedAt: nextUpdatedAt } = await Descriptions.findById(pokemon1Id);

            expect(nextUpdatedAt.valueOf()).toBeGreaterThan(prevUpdatedAt.valueOf());
        });
    });

    describe('Unpublish processed row object', () => {
        it('should return proper response', async () => {
            const response = await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`)
                .send({ published: false });
            const unpublishedRow = await ProcessedRowsModel.findById(getId(pokemon1RowObject1));

            expect(response).toMatchSnapshot();
            expect(unpublishedRow).not.toBeNull();
        });

        it('should update unpublished key in db', async () => {
            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`)
                .send({ published: false });

            expect(await ProcessedRowsModel.findById(getId(pokemon1RowObject1)))
                .toHaveProperty('published', false);
        });

        it('should remove unpublished row\'s cached field values if there is no more row objects with such values', async () => {
            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`)
                .send({ published: false });

            const cachedValue = await cached_values.findById(getId(pokemon1CachedValues));

            expect(cachedValue.limitedUniqValsByColName[pokemonNoField])
                .not.toContain(pokemon1RowObject1.rowParams[pokemonNoField]);
        });

        it('should remove unpublished row\'s cached field values if there are only unpublished rows with those values', async () => {
            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject9)}`)
                .send({ published: false });

            const cachedValue = await cached_values.findById(getId(pokemon1CachedValues));

            expect(cachedValue.limitedUniqValsByColName[pokemonType1Field])
                .not.toContain(pokemon1RowObject9.rowParams[pokemonType1Field]);
        });

        it('should not remove unpublished row\'s cached field values if there are still objects with those values', async () => {
            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`)
                .send({ published: false });

            const cachedValue = await cached_values.findById(getId(pokemon1CachedValues));

            expect(cachedValue.limitedUniqValsByColName[pokemonType1Field])
                .toContain(pokemon1RowObject1.rowParams[pokemonType1Field]);
        });

        it('should update dataset modify date', async () => {
            const { updatedAt: prevUpdatedAt } = await Descriptions.findById(pokemon1Id);

            await appRequest
                .put(`/api/dataset/content/${pokemon1Id}/${getId(pokemon1RowObject1)}`)
                .send({ published: false });

            const { updatedAt: nextUpdatedAt } = await Descriptions.findById(pokemon1Id);

            expect(nextUpdatedAt.valueOf()).toBeGreaterThan(prevUpdatedAt.valueOf());
        });
    });
});
