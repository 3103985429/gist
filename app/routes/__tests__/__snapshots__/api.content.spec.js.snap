// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content API Create processed row object should return proper response 1`] = `
Object {
  "body": Object {
    "created": true,
    "published": true,
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/5/5d/010Caterpie.png/250px-010Caterpie.png",
      "Max CP": 446,
      "Max HP": 83,
      "Name": "Caterpie",
      "Pokemon No_": 10,
      "Type 1": "Bug",
      "Type 2": "None",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
  },
  "statusCode": 200,
}
`;

exports[`Content API Publish processed row object should return proper response 1`] = `
Object {
  "body": Object {
    "pKey": "14",
    "published": true,
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/3/39/007Squirtle.png/250px-007Squirtle.png",
      "Max CP": 1015,
      "Max HP": 81,
      "Name": "Squirtle",
      "Pokemon No_": 7,
      "Type 1": "Water",
      "Type 2": "None",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
  },
  "statusCode": 200,
}
`;

exports[`Content API Remove processed row object should return proper response 1`] = `
Object {
  "body": Object {
    "success": true,
  },
  "statusCode": 200,
}
`;

exports[`Content API Unpublish processed row object should return proper response 1`] = `
Object {
  "body": Object {
    "pKey": "2",
    "published": false,
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/2/21/001Bulbasaur.png/250px-001Bulbasaur.png",
      "Max CP": 1079,
      "Max HP": 83,
      "Name": "Bulbasaur",
      "Pokemon No_": 1,
      "Type 1": "Grass",
      "Type 2": "Poison",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
  },
  "statusCode": 200,
}
`;

exports[`Content API Update row object's params should return proper response 1`] = `
Object {
  "body": Object {
    "pKey": "2",
    "published": true,
    "rowParams": Object {
      "Image URL": "http://cdn.bulbagarden.net/upload/thumb/2/21/001Bulbasaur.png/250px-001Bulbasaur.png",
      "Max CP": 1079,
      "Max HP": 83,
      "Name": "Bulbasaur",
      "Pokemon No_": 12,
      "Type 1": "Grass",
      "Type 2": "Poison",
    },
    "srcDocPKey": "5a42c26e303770207ce8f83b",
  },
  "statusCode": 200,
}
`;
